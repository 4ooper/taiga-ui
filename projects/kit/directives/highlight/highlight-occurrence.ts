export interface TuiHighlightOccurrence {
    index: number;
    length: number;
}
