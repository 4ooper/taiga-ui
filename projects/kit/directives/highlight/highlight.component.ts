import {ChangeDetectionStrategy, Component} from '@angular/core';

@Component({
    standalone: true,
    selector: 'mark[tuiHighlightMark]',
    template: '',
    styleUrls: ['./highlight.style.less'],
    changeDetection: ChangeDetectionStrategy.OnPush,
})
export class TuiHighlightComponent {}
