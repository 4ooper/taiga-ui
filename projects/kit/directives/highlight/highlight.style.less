:host {
    background: var(--tui-highlight-background, var(--tui-selection));
    border-radius: var(--tui-highlight-radius, 0);
    color: var(--tui-highlight-color, currentColor);
}
