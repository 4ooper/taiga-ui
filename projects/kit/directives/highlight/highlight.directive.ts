import {DOCUMENT, isPlatformBrowser} from '@angular/common';
import type {OnChanges} from '@angular/core';
import {
    booleanAttribute,
    Directive,
    ElementRef,
    EnvironmentInjector,
    inject,
    Input,
    PLATFORM_ID,
} from '@angular/core';
import {takeUntilDestroyed} from '@angular/core/rxjs-interop';
import type {TuiArrayOrValue} from '@taiga-ui/cdk';
import {svgNodeFilter, tuiIsNumber, tuiToArray} from '@taiga-ui/cdk';
import {TuiToRegexpPipe} from '@taiga-ui/kit/pipes';
import {mergeAll, Subject, switchMap, tap} from 'rxjs';

import {TuiHighlight} from './highlight';
import type {TuiHighlightOccurrence} from './highlight-occurrence';

@Directive({
    standalone: true,
    selector: '[tuiHighlight]',
})
export class TuiHighlightDirective implements OnChanges {
    private patterns: readonly RegExp[] = [];
    private readonly highlight$ = new Subject<TuiHighlight>();
    private readonly resetHighlights$ = new Subject<void>();
    private readonly toRegexpPipe = inject(TuiToRegexpPipe);
    private readonly environmentInjector = inject(EnvironmentInjector);
    private readonly doc = inject(DOCUMENT);
    private readonly isBrowser = isPlatformBrowser(inject(PLATFORM_ID));
    private readonly treeWalker = this.doc.createTreeWalker(
        inject(ElementRef).nativeElement,
        NodeFilter.SHOW_TEXT,
        svgNodeFilter,
    );

    @Input({
        transform: booleanAttribute,
    })
    public tuiHighlightMultiOccurrences = false;

    constructor() {
        this.resetHighlights$
            .pipe(
                switchMap(() =>
                    this.highlight$.pipe(
                        mergeAll(),
                        tap(node => {
                            this.treeWalker.currentNode = node;
                        }),
                    ),
                ),
                takeUntilDestroyed(),
            )
            .subscribe();
    }

    @Input()
    public set tuiHighlight(value: TuiArrayOrValue<RegExp | string>) {
        this.patterns = tuiToArray(value).map(item => {
            if (item instanceof RegExp) {
                // Only global regexp's can be used in String.prototype.mathAll method
                if (!item.global) {
                    return new RegExp(item.source, `${item.flags}g`);
                }

                return item;
            }

            return this.toRegexpPipe.transform(item, 'gi');
        });
    }

    public get tuiHighlight(): readonly RegExp[] {
        return this.patterns;
    }

    public ngOnChanges(): void {
        if (!this.isBrowser) {
            return;
        }

        queueMicrotask(() => this.createHighlights());
    }

    private createHighlights(): void {
        this.resetHighlights$.next();

        for (const node of this.getNodes()) {
            const occurrence = this.getFirstOccurrence(node.nodeValue);

            if (occurrence) {
                this.highlight$.next(
                    new TuiHighlight(
                        this.environmentInjector,
                        this.createRange(node, occurrence),
                    ),
                );

                if (!this.tuiHighlightMultiOccurrences) {
                    return;
                }
            }
        }
    }

    private createRange(node: Node, {index, length}: TuiHighlightOccurrence): Range {
        const range = this.doc.createRange();

        range.setStart(node, index);
        range.setEnd(node, index + length);

        return range;
    }

    private getFirstOccurrence(source: string | null): TuiHighlightOccurrence | null {
        if (!source) {
            return null;
        }

        let lastApprovedOccurrence: TuiHighlightOccurrence | null = null;

        for (const item of this.tuiHighlight) {
            const [match] = source.matchAll(item);

            if (
                match &&
                tuiIsNumber(match.index) &&
                match[0].length &&
                (!lastApprovedOccurrence || lastApprovedOccurrence.index > match.index)
            ) {
                lastApprovedOccurrence = {
                    index: match.index,
                    length: match[0].length,
                };
            }
        }

        return lastApprovedOccurrence;
    }

    private *getNodes(): Generator<Node> {
        this.treeWalker.currentNode = this.treeWalker.root;

        while (this.treeWalker.nextNode()) {
            yield this.treeWalker.currentNode;
        }
    }
}
