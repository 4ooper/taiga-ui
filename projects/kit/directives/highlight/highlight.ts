import type {ElementRef, EnvironmentInjector} from '@angular/core';
import {createComponent} from '@angular/core';
import {Observable} from 'rxjs';

import {TuiHighlightComponent} from './highlight.component';

export class TuiHighlight extends Observable<Node> {
    constructor(environmentInjector: EnvironmentInjector, range: Range) {
        super(observer => {
            const component = createComponent(TuiHighlightComponent, {
                environmentInjector,
            });
            const {nativeElement} = component.location as ElementRef<Node>;

            range.surroundContents(nativeElement);
            observer.next(nativeElement.firstChild!);

            return () => {
                const parentNode = nativeElement.parentNode;

                if (parentNode) {
                    while (nativeElement.firstChild) {
                        parentNode.insertBefore(nativeElement.firstChild, nativeElement);
                    }

                    parentNode.removeChild(nativeElement);

                    const firstNode = parentNode.firstChild;

                    if (firstNode instanceof Text) {
                        while (firstNode.nextSibling) {
                            if (firstNode.nextSibling instanceof Text) {
                                firstNode.data += firstNode.nextSibling.data;
                                parentNode.removeChild(firstNode.nextSibling);
                            } else {
                                break;
                            }
                        }
                    }
                }

                component.destroy();
            };
        });
    }
}
