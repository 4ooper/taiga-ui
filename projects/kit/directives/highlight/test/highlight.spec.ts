import {EnvironmentInjector} from '@angular/core';
import {TestBed} from '@angular/core/testing';

import {TuiHighlight} from '../highlight';

describe('TuiHighlight', () => {
    let environmentInjector: EnvironmentInjector;

    beforeEach(() => {
        environmentInjector = TestBed.inject(EnvironmentInjector);
    });

    it('should create a component and insert it into a DOM Range', () => {
        const rootElement = document.createElement('div');
        const textNode = document.createTextNode('Hello, world!');
        const range = new Range();

        range.setStart(textNode, 0);
        range.setEnd(textNode, 5);

        rootElement.appendChild(textNode);

        const highlight = new TuiHighlight(environmentInjector, range);
        const subscribeFn = jest.fn();

        highlight.subscribe(subscribeFn);

        expect(subscribeFn).toHaveBeenCalledWith(document.createTextNode('Hello'));
        expect(rootElement).toMatchSnapshot();
    });

    it('should clean up the DOM and component on unsubscribe', () => {
        const rootElement = document.createElement('div');
        const textNode = document.createTextNode('Hello, world!');
        const range = new Range();

        range.setStart(textNode, 0);
        range.setEnd(textNode, 5);

        rootElement.appendChild(textNode);

        const highlight = new TuiHighlight(environmentInjector, range);

        highlight.subscribe().unsubscribe();

        expect(rootElement).toMatchSnapshot();
    });
});
