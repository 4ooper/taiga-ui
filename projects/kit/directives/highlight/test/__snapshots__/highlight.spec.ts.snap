// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TuiHighlight should clean up the DOM and component on unsubscribe 1`] = `
<div>
  Hello, world!
</div>
`;

exports[`TuiHighlight should create a component and insert it into a DOM Range 1`] = `
<div>
  
  <mark
    tuihighlightmark=""
  >
    Hello
  </mark>
  , world!
</div>
`;
